# Importing the required libraries

In [11]:
# Import required libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.layers import Input


# Load the data

In [12]:
# Load the data
data = pd.read_csv('https://cocl.us/concrete_data')

data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

# Split the data into predictors and target

In [13]:
X = data.drop(columns=['Strength'])
y = data['Strength']

# Normalize the data

In [14]:
from sklearn.preprocessing import StandardScaler

# Normalize the data
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

# Build a neural network model

In [15]:
#Build a baseline model 
def build_model():
    model = Sequential([
        Input(shape=(X.shape[1],)),  # Define the input shape using an Input layer
        Dense(10, activation='relu'),  # Hidden layer with 10 nodes and ReLU activation
        Dense(1)  # Output layer with 1 node (regression task)
    ])
    model.compile(optimizer=Adam(), loss='mean_squared_error')
    return model

# Test the model

In [16]:
# Train and evaluate the model
def train_and_evaluate():
    X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.3, random_state=np.random.randint(10000))
    model = build_model()
    model.fit(X_train, y_train, epochs=100, batch_size=10, verbose=0)  # Set verbose to 0 to suppress output
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    return mse


# Train and evaluate the model

In [17]:
mse_list = []
for i in range(100):
    mse = train_and_evaluate()
    mse_list.append(mse)
    print(f'Mean Squared Error {i+1}: {mse}')
    

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 1: 67.74447454662388
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 2: 68.24698314949885
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 3: 51.3347362879281
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 4: 72.96428899384907
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 5: 71.89275610302984
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 6: 73.82042747095365
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 7: 69.42621008310095
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 8: 80.38622215347576
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 9: 58.24970648223607
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 10: 85.68482350674414
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 
Mean Squared Error 11: 79.17683020457892
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 12: 74.8568780277139
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step 

# Calculate the mean and standard deviation of the mean squared errors

In [18]:

# Calculate the mean and standard deviation of the mean squared errors
mse_list = np.array(mse_list)
mean = mse_list.mean()
std = mse_list.std()
print(f'Mean of mean squared errors: {mean}')
print(f'Standard deviation of mean squared errors: {std}')


Mean of mean squared errors: 77.32465413094039
Standard deviation of mean squared errors: 15.31779256987466


# How does the mean of the mean squared errors compare to that from Step A?

## In step B, the mean of the mean squared errors is 130.05 and the standard deviation is 11.77.In this step the mean of the mean squared errors is 77.32 and the standard deviation is 15.3177.32 is less than 130.05 and 15.31 is more than 11.77. So, the model is better than the model in step B. 
#%